### PR TITLE
Idserver update

### PR DESCRIPTION
**Pull request by rockfruit**
This builds on (and supersedes) https://github.com/senaite/bika.lims/pull/310.

Includes:
- [x] https://github.com/bikalims/bika.lims/pull/2210
- [x] https://github.com/bikalims/bika.lims/pull/2294
- [x] https://github.com/bikalims/bika.lims/pull/2273

### CODE REVIEW

File: bika/lims/idserver.py [hunk 1]
```diff
@@ -43,154 +41,125 @@ def idserver_generate_id(context, prefix, batch_size = None):
...
-        new_id = str(idserver_generate_id(context, npt + "-"))
-        return ('%s' + separator + '%s') % (npt, new_id)
+    if portal_type == '':
+        portal_type = context.portal_type
 
+    def getLastCounter(context, config):
+        if config.get('counter_type', '') == 'backreference':
+            relationship = config['counter_reference']
+            backrefs = get_backreferences(context, relationship)
+            return len(backrefs) - 1
```
Comment (xispa) [inline bika/lims/idserver.py]: IDServer doc test, while creating Analysis Requests, fail:
```
Differences (ndiff with -expected +actual):
    - <...water17-0001-R1>
    + <AnalysisRequest at /plone/clients/client-1/water17-0001-R0>
```

`backrefs` length is always 0 and the function ends up returning -1. I've been walking through `uidreferencefield`, checking `get_backreferences`, `_set_backreferences`, etc. before and after the AR's `processForm` in `create_analysisrequest` and I always get valid backreferences from Sample. The weird thing is that I've tested IDServer running the instance and all works perfect.... so I started to think about `transaction.savepoint()`, etc... but w/o success. The UIDReferenceField doc test works too, so I'm a bit lost here........ I'll keep digging.

Comment (xispa) [reply bika/lims/idserver.py]: Calling `get_backreferences(obj.getSample())` after this:

https://github.com/rockfruit/bika.lims/blob/1f0fc0fefa0e3719e46e2510708f8400cf851190/bika/lims/idserver.py#L176-L177

Investigating now what the hell is going on with this savepoint...

Comment (xispa) [reply bika/lims/idserver.py]: From @ramonski in gitter
I think the issue regarding the backreferences in IDServer is located here: 
https://github.com/senaite/bika.lims/blob/senaite-integration/bika/lims/browser/fields/uidreferencefield.py#L144

This is a simple list type and Zope does not recognize for these types when the data has changed.

https://docs.plone.org/develop/plone/persistency/persistent.html#lists-and-dictionaries

Fixed here:
https://github.com/senaite/bika.lims/commit/93cd343a2bab4b76a32b210ef8ee28bf30d773a1

AFAIK mutable objects are read/written in one transaction completely, becuase the ZODB does not recognize that these have changed

Therfore the `_p_changed = True` dance

**Comment by rockfruit** [comment]
Ran 1 tests with 0 failures and 0 errors in 7.588 seconds.

**Pull request merged**